Enable editing of links and homepage URLs

brenetic opened the pull request:
- Added ability to edit homepage URL's
- Enable the editing of links
# Screenshots

<img width="1483" alt="screen shot 2016-06-03 at 10 57 59" src="https://cloud.githubusercontent.com/assets/3466862/15775602/ab76a1ce-297a-11e6-8ea9-b062c2b7e358.png">

<img width="1484" alt="screen shot 2016-06-03 at 11 01 01" src="https://cloud.githubusercontent.com/assets/3466862/15775607/aeddc518-297a-11e6-9b67-f6f85de16ae0.png">

<img width="1464" alt="screen shot 2016-06-03 at 10 55 45" src="https://cloud.githubusercontent.com/assets/3466862/15775610/b3b4f0d4-297a-11e6-8122-61ff4f29f209.png">

<img width="1481" alt="screen shot 2016-06-03 at 10 55 32" src="https://cloud.githubusercontent.com/assets/3466862/15775618/bf8c7210-297a-11e6-9373-00979dae9d7b.png">

<img width="1454" alt="screen shot 2016-06-03 at 10 56 00" src="https://cloud.githubusercontent.com/assets/3466862/15775623/c4db748c-297a-11e6-95d5-8ed30d86aa01.png">

[Trello editing links](https://trello.com/c/qtpaquan/389-enable-editing-of-links)
[Trello editing homepage url](https://trello.com/c/DfDlIIRc/390-make-the-homepage-url-for-each-local-authority-editable)

Mobbed by: @emmabeynon @sihugh @klssmith @brenetic @issyl0

brenetic committed:
Add slug to interactions

We add the slug field to the Interaction model so that we can use them
as human readable URLs

brenetic committed:
Add edit button and enable editing of links

brenetic committed:
Add save link functionality

Displays flash messages on success and failure.

brenetic committed:
Add breadcrumb to edit links page

brenetic committed:
Add partial for local authority details

brenetic committed:
Add the ability to edit homepage URLs for local authorities

issyl0 commented on app/controllers/links_controller.rb:
```diff
@@ -0,0 +1,43 @@
+    @link = Link.get_link(params)
+    if @link.nil?
+      @link = Link.new
+      @link.url = 'n/a'
```
This could be shortened to `@link = Link.new(url: 'n/a')`.

davidbasalla replied:
I'm wondering whether we should leave `@link.url` as `nil` here and move the logic for displaying 'n/a' to the view or a helper like LinksHelper? To me showing 'n/a' is more a question of what we display when we get a `nil` link. Changing the value of `@link.url` to 'n/a' technically means that it's not `nil` anymore. Not sure if I'm making sense. What do you think?

davidbasalla replied:
Also I haven't tried it yet but could we use a kind of `first_or_create`/`find_or_create_by` pattern here?

sihugh replied:
I think we were happy with not persisting links that don't have a real URL in them, but it sounds good.  I think moving the `if nil?` use `n/a` to a helper is a good idea.

davidbasalla replied:
One more question on this, do we even need to set this to be 'n/a'? Would it be okay just to leave `@link.url` as `nil`? The field in the form would then just be empty which still indicates that there is no url for this link.

sihugh replied:
Good question!  We should ask the rest of the hive brain 🐝

h-lame replied:
+1 for keeping the 'n/a' out of the controller

Further to @davidbasalla's comment about `first_or_create` it might be interesting to make `Link.get_link` _always_ return a `Link` instance.  In the case where the provided arguments allow us to fetch something from the DB the behaviour remains the same, but when those args don't resolved to a row in the DB we could build an unsaved `Link` instance and set it up with appropriate `LocalAuthority` and `ServiceInteraction` associations (e.g. what we do in the `create` action) so that we _could_ save it once we've filled in the url.

I don't know that this is a great idea, but I think it's interesting, and it's a pattern I've used before.

sihugh replied:
Thanks for the comments. We have addressed this now. Will update in a bit

issyl0 commented on app/controllers/links_controller.rb:
```diff
@@ -0,0 +1,43 @@
+
+private
+
+  def validate_and_save(link)
```
ActiveRecord checks validations by default on `save` - http://guides.rubyonrails.org/active_record_validations.html#when-does-validation-happen-questionmark - do we need this method?

sihugh replied:
The main bit of this wasn't firing the validations so much as the redirections and flash messages.  I expect these will change a bit after the show and tell.  We could compress the `if link.valid?` and `link.save!` lines into just `if link.save` though.

davidbasalla commented on spec/controllers/links_controller_spec.rb:
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe LinksController, type: :controller do
```
Should we add some tests for the controller actions here?

sihugh replied:
We probably should.  We have got feature tests that cover this though 👍

h-lame replied:
FWIW: controller tests are going away in rails 5 (http://blog.bigbinary.com/2016/04/19/changes-to-test-controllers-in-rails-5.html explains the changes in detail - but not the rationale.  rspec is going to do the work that means you'll still be able to write them with rspec so when/if we upgrade we should be able to keep them, but the writing is on the wall that directly testing your controller is a no-longer-the-rails-way kinda thing to do so it might get harder and harder to do.

This is not to say that we shouldn't test our controllers, but that the specifics of how rails told you to test your controllers in the past (checking which template was rendered, checking which variables are assigned to share with the view, etc..) are not the way to do it.  Instead rails 5 suggests you test controllers with integration tests (feature specs in rspec parlance) where you don't reach inside to check the internals, but you can test the "full-stack" by checking what's on the rendered views.

So, all this to say, we have feature specs already so unless there's some complex branching route through the controller that we haven't covered in those, it's probably better to not introduce controller specs.

davidbasalla replied:
Thanks for sharing, good to know!

h-lame commented on app/views/links/edit.html.erb:
```diff
@@ -0,0 +1,21 @@
+
+<%= label_tag 'link', 'URL' %>
+
+<%= form_for @link, url: {action: "update"} do |f| %>
```
I don't think this `url: {action: "update"}` is doing anything as when `@link` is unsaved we end up in the `create` action, and when it is saved we end up in the `update` action.  This is the default behaviour of `form_for` so our `url` param isn't doing anything (and acts to confuse the reader as they'll think we always end up in the `update` action which isn't the case).

That said, if our intention is to always go through the `update` action regardless of it being a saved or unsaved (not a terrible idea if we can do something with @davidbasalla's idea of using some kind of `first_or_intialize` in / around our `Link.get_link` call) then this is the right sort of thing to do, but it's not working - maybe worth exploring this a little bit more.

h-lame commented on app/controllers/links_controller.rb:
```diff
@@ -0,0 +1,43 @@
+      redirect_to local_authority_service_interactions_path(service_slug: params[:service_slug])
+    else
+      flash[:danger] = "Please enter a valid link."
+      redirect_to action: "edit"
```
Because we redirect here we lose a load of information about what the user has done.
1. We lose the url that they entered - so when we render the form again we can only show them the url from the db (or "n/a")
2. We lose the `link` instance that we called `.valid?` on which means we lose the error messages on there - when rails renders a form using `form_for` the form helpers will look at the object we passed to that to decide if the input elements should be given error classes to style them differently.

This is because a redirect causes a whole new http request/response to happen and the `link` instance and user entered params only exist for the request/response they were created in.  If instead of redirecting we asked to render the edit form again (`render :edit`) we'd keep the `link` instance so that rendering the form could use the validation errors and the url that the user had entered already.  Note that the form wants to use `@link` but our `create` action only uses `link`...

h-lame commented on app/controllers/links_controller.rb:
```diff
@@ -0,0 +1,43 @@
+    link.local_authority = LocalAuthority.find_by(slug: params[:local_authority_slug])
+    link.service = Service.find_by(slug: params[:service_slug])
+    link.interaction = Interaction.find_by(slug: params[:interaction_slug])
+    link.service_interaction = ServiceInteraction.find_by(service: link.service, interaction: link.interaction)
```
We don't need to set `service`, `interaction`, and `service_interaction`.  If we set `service_interaction` then we'll get `service` and `interaction` for "free" because our `service` and `interaction` associations come _through_ the `service_interaction` association.

h-lame commented on app/controllers/local_authorities_controller.rb:
```diff
@@ -2,4 +2,27 @@ class LocalAuthoritiesController < ApplicationController
+
+private
+
+  def validate_and_save(authority)
```
What's been said already about `valid?` and `save!` and `redirect_to` in the `LinksController` applies here too.

h-lame commented on app/models/link.rb:
```diff
@@ -14,4 +14,12 @@ class Link < ActiveRecord::Base
+      services: { slug: params[:service_slug] },
+      interactions: { slug: params[:interaction_slug] }
+    )
+  end
```
@brenetic said that when doing this as a scope it would return loads of `Link` instances instead of just one.  The reason is that scopes _always_ return an array (techincally it's an `ActiveRecord::Relation` instance - which is what allows us to chain them together) - this is even if you've used a method like `.first` or `.find_by` which usually returns a single instance.  When you write it as a class method then there's none of the "magic" scope stuff happening to make sure we always return something chainable so in this case we would always get a single instance (or nil if one can't be found).

brenetic committed:
Add highlighting to field with errors

brenetic committed:
Amend logic to display text for absent link

brenetic committed:
Show previous text if link validation fails

brenetic committed:
Show previous text when invalid homepage entered

brenetic committed:
Add simple link controller test

Our controller is mostly being exercised via the feature tests.

brenetic committed:
Change local authority edit homepage button wording

brenetic committed:
Show 'No link' when none is available

brenetic committed:
Highlight saved row on interactions page

brenetic committed:
Display full homepage url for local authority

brenetic committed:
Add example url to link edit pages

brenetic committed:
Changed position of flash notices

brenetic committed:
Fix edit homepage link cancel button bug

Also fixed missing breadcrumb on the same page.

brenetic committed:
Align and style LGIL/LGSL codes

brenetic committed:
Refactor Link model

When a new link is created a link object is returned with all necessary
services and interactions already set.

Commit:
Fix the create/update link

- We'll need to do some overriding of `to_param` to make Rails use
  `slug` by default, and this isn't the most elegant solution as we have
  to specify explicit `url`s, but it works for now and the overriding
  work (thanks, Murray!) is for another story/PR.
